### PR TITLE
Fixes for firewall issues

### LINKED ISSUE #2768: New Firewall Fields Stay in Error State
If user wants to open a port with invalid parameters, Kura colours the invalid fields in red. If user then cancels the operation and clicks "New" again, all the invalid fields from before are still marked red.

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall, click "new" and enter all or some parameters that are invalid
3. Click Cancel and then "new" again - observe the fields that are marked red

**Expected behavior**
If user cancels the operation and then again clicks on "new" button, all the fields should be reset. 

**Screenshot No. 1:**
Screenshot 2020-04-06 at 11 44 04

**Screenshot No. 2:**
Screenshot 2020-04-06 at 11 47 36

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Additional info: /

### LINKED ISSUE #2789: "Internal Port" and "External Port" in Port Forwarding not Marked as Mandatory
If user wants to set port forwarding, there are three fields that are marked as mandatory and those are:
 - input interface
 - output interface
 - Lan address

But when user clicks Submit after populating those fileds, the Internal port and External port are put in error state, because they were not populated (they are not marked as mandatory). 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall -> Port Forwarding
3. Enter all the mandatory parameters (valid)
4. Hit Submit and observe the two fields in error state

**Expected behavior**
If the two fields are mandatory, they should be "equiped" with the asterix symbol ("*") as all amndatory fields. If they are not, user should be able to submit the settings without specifying those. 

**Screenshot:**
Screenshot 2020-04-09 at 07 40 31

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.1.0

### LINKED ISSUE #2834: "Duplicate Entry" Error When Submitting Empty Open Port Information
If user opens "new" dialogue box in Firewall and clicks "Submit" without specifying a thing, Kura returns "duplicate entry" error, although there is no duplicate entry. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall, click "new"
3. Without entering anything click on Submit - observe the error.

**Expected behavior**
Kura should not report a "Duplicate error" if there are no duplicates in the Firewall list.

**Screenshot:**
Screenshot 2020-05-05 at 12 45 17

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.2.0_SNAPSHOT ([Link to the specific Kura version -from May 4th 2020](https://ci.eclipse.org/kura/job/kura-develop/lastSuccessfulBuild/artifact/kura/distrib/target/kura_4.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb))

### LINKED ISSUE #2844: "Open Ports" Changes Lost if View Changed
If user has some unsaved changes in Open Ports in Firewall (e.g. some added port without clicking Apply), if the view changes to e.g."Port Forwarding", all the changes are lost. 
This can be a real problem if user has entered a lot of open ports and then accidentally clicks on something else. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall -> Open ports
3. Create at least one calid open port and without clicking "Apply" click on "Port Forwarding"
4. Return to Open ports - observe the list

**Expected behavior**
If user has unsaved changes in "Open Ports" section, Kura should ask him if he wants to disscard the changes or not before proceeding. 

**Screenshot:**
/

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.2.0_SNAPSHOT ([Link to the specific Kura version -from May 4th 2020](https://ci.eclipse.org/kura/job/kura-develop/lastSuccessfulBuild/artifact/kura/distrib/target/kura_4.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb))

### LINKED ISSUE #2845: Multiple Colons Allowed in Port and Port Range Fields in Firewall
If user enters e.g. 7100::::::7200 port range in Port range fields in Firewall, the fields are not in error state when Submit is clicked. Kura reports an error when trying to apply changes. 
I think that Kura should mark the field in red earlier (whn Submit is clicked). 

There is also another poblem (Screenshot No. 3) - because Kura does not recognize multiple semicolons as a bug, we can add as much as we want to and when this port entry is clicked to be edited, the name is not properly dispalyed. 

EDIT: This happens in all fields, not only in Open Ports tab, but also in Port forwarding!

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall -> Open Ports 
3. Click New
4. Add all the mandatory parameters and also port range as "1:::::::3" (multiple colons)
5. Click Submit and Apply

**Expected behavior**
This should not be possible, only one colon should be allowed. 

**Screenshot No. 1:**
Screenshot 2020-05-06 at 10 13 15

**Screenshto No. 2:**
Screenshot 2020-05-06 at 10 10 44

**Screenshot No. 3:**
Screenshot 2020-05-06 at 10 11 10

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.2.0_SNAPSHOT ([Link to the specific Kura version -from May 4th 2020](https://ci.eclipse.org/kura/job/kura-develop/lastSuccessfulBuild/artifact/kura/distrib/target/kura_4.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb))

### LINKED ISSUE #2846: Port Forwarding Entries Not Properly Deleted
If user has at least two entries in the Port Forwarding section of Firewall service, selects the second one, click on Delete and confirms, the entry is not deleted - it is still in the list (see the screenshots bellow). 

**Testflow:**
1. Login to Kura as admin/admin
2. Go Firewall -> Port Forwarding
3. Enter two different Ports to be forwarded
4. Apply changes
5. Select second one and click on Delete. Confirm the deletion
6. Observe the list (Screenshot No. 2)

**Expected behavior**
If user deletes an entry in port forwarding list, it should disappear from the list. 

**Screenshot No. 1:**
Screenshot 2020-05-06 at 09 41 44

**Screenshot No. 2:**
Screenshot 2020-05-06 at 10 57 53

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.2.0_SNAPSHOT ([Link to the specific Kura version -from May 4th 2020](https://ci.eclipse.org/kura/job/kura-develop/lastSuccessfulBuild/artifact/kura/distrib/target/kura_4.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb))

### LINKED ISSUE #2847: "Port Forwarding" Changes Lost if View Changed
If user enters a specific port to be forwarded (all valid parameters) and then changes views to e.g. "IP Forwarding/Masquerading" and then returns to "Port Forwarding", all the unsaved changes are saved. 

**Testflow:**
1. Login to Kura as admin/admin
2. Go to Firewall -> port Forwarding 
3. Add an entry and without clicking on Apply change view to "IP forwarding/Masquerading"
4. Go back to Firewall -> port forwarding 
5. Observe the missing entry
It is true that as user you did not apply changes, but Kura should at least notify you about this. 

**Expected behavior**
User should be prompted if the changes should be saved or not prior to changing the views in Firewall. 

**Screenshot No. 1:**
Screenshot 2020-05-06 at 11 23 28

**Screenshot No. 2:**
Screenshot 2020-05-06 at 11 23 50

Target Environment (please complete the following information):
Board: Raspberry Pi 3
OS version: Linux raspberrypi 4.19.75-v7+ # 1270 SMP Tue Sep 24 18:45:11 BST 2019 armv7l GNU/Linux
Kura version: 4.2.0_SNAPSHOT ([Link to the specific Kura version -from May 4th 2020](https://ci.eclipse.org/kura/job/kura-develop/lastSuccessfulBuild/artifact/kura/distrib/target/kura_4.2.0-SNAPSHOT_raspberry-pi-2-3_installer.deb))

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
This PR tries to fix the pending firewall issues.

**Related Issue:** This PR 
closes #2768 
closes #2789 
closes #2834 
closes #2844 
closes #2845 
closes #2846 
closes #2847

**Description of the solution adopted:** The PR adds a set of checks on the PortForwarding modal triggered at submit button press; the IpForwarding tab is improved to correctly remove the delete rows from the table; additional check are added to avoid the submit of an empty entry; a modal is added when the user moves from a (dirty) tab to another.

### CODE REVIEW

**Comment by pierantoniomerlino** [comment]
@LeoNerdoG This should fix the remaining pending issues about firewall. Could you please take a look?

**Comment by LeoNerdoG** [comment]
Hey @pierantoniomerlino,
I have checked all the issues mentioned in the PR's description:
 - 2789: OK
 - 2834: OK
 - 2844: OK
 - 2845: 
    - **In "Open Ports"**: Error is reported when user clicks Apply (but should relly be when trying to click Submit - except if you wanted this?)
Screenshot: 
Screenshot 2020-06-05 at 12 22 46
     - **In "Port Forwarding"**: Changes still saved after closing error and clicking "yes" in window "Are you sure to discrad you changes" and returning back to Port Forwarding
Screenshot 2020-06-05 at 12 42 46

 - 2846: OK
 - 2847: OK

EDIT: 
As you can see from the tag bellow, this PR also fixes issue 2768.

**Comment by pierantoniomerlino** [comment]
@LeoNerdoG I'd have fixed the multiple colons in open ports and port forwarding tabs. Could you please take a look?

**Comment by LeoNerdoG** [comment]
@pierantoniomerlino I am confirming that all of the issue above (+2768) are properly fixed and this can be merged into develop branch.

**Comment by MMaiero** [comment]
I've tried the changes.
I've seen that the message in firewall when moving out of the current tab is properly displayed. But, if I choose to leave the page and move back in, the uncommitted change is still there. Is that expected?

**Comment by LeoNerdoG** [comment]
I think this is not the desired behaviour. Every time user willingly presses "discard changes", all changes should be lost - @pierantoniomerlino can you fix this also?

**Comment by pierantoniomerlino** [comment]
@MMaiero It happens only for the PortForwarding and IpForwarding tabs...

**Comment by pierantoniomerlino** [comment]
@MMaiero @LeoNerdoG I fixed the Port and IpForwarding behavior. Moreover, now when the Firewall service is clicked, the OpenPorts is displayed.

**Comment by LeoNerdoG** [comment]
hey @pierantoniomerlino,
I have checked the latest changes and now it seems that everything is good. 😄 
I think this can be merged.

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/Tab.java [hunk 1]
```diff
@@ -22,6 +22,8 @@
 
     public void refresh();
 
+    public void clear();
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/Tab.java:25]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/ProfileTabUi.java [hunk 2]
```diff
@@ -151,4 +151,8 @@ public void onSuccess(ArrayList<GwtGroupedNVPair> result) {
         });
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/ProfileTabUi.java:154]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/SystemPropertiesTabUi.java [hunk 3]
```diff
@@ -146,4 +146,8 @@ public void onSuccess(ArrayList<GwtGroupedNVPair> result) {
         });
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/SystemPropertiesTabUi.java:149]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/ThreadsTabUi.java [hunk 4]
```diff
@@ -142,4 +142,9 @@ public void onSuccess(ArrayList<GwtGroupedNVPair> result) {
             }
         });
     }
+
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/device/ThreadsTabUi.java:146]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/firewall/FirewallPanelUtils.java [hunk 5]
```diff
@@ -40,11 +40,11 @@ private static boolean checkPort(String port) {
...
 
     public static boolean checkPortRegex(String ports) {
-        boolean isPortRegex = false;
-        if (ports.trim().matches(FieldType.PORT_RANGE.getRegex()) || ports.trim().matches(FieldType.PORT.getRegex())) {
-            isPortRegex = true;
-        }
-        return isPortRegex;
+        return ports.trim().matches(FieldType.PORT.getRegex());
+    }
+
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/firewall/FirewallPanelUtils.java:45]: Unnecessary empty line?

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/firewall/NatTabUi.java [hunk 6]
```diff
@@ -446,10 +447,12 @@ public void onDelete() {
         if (selection != null) {
             this.alertDialog.show(MSGS.firewallNatDeleteConfirmation(selection.getInInterface()), () -> {
                 NatTabUi.this.natDataProvider.getList().remove(selection);
-                refreshTable();
                 NatTabUi.this.buttonBar.setApplyResetButtonsDirty(true);
                 NatTabUi.this.buttonBar.setEditDeleteButtonsDirty(false);
                 NatTabUi.this.selectionModel.setSelected(selection, false);
+                setVisibility();
+                refreshTable();
+
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/firewall/NatTabUi.java]: Unnecessary empty line?

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabDhcpNatUi.java [hunk 7]
```diff
@@ -213,6 +213,11 @@ public void getUpdatedNetInterface(GwtNetInterfaceConfig updatedNetIf) {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabDhcpNatUi.java:216]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabHardwareUi.java [hunk 8]
```diff
@@ -128,6 +128,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabHardwareUi.java:131]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemAntennaUi.java [hunk 9]
```diff
@@ -116,6 +116,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemAntennaUi.java:119]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemGpsUi.java [hunk 10]
```diff
@@ -121,6 +121,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemGpsUi.java:124]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemUi.java [hunk 11]
```diff
@@ -345,6 +345,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabModemUi.java:348]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabTcpIpUi.java [hunk 12]
```diff
@@ -338,6 +338,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabTcpIpUi.java:342]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabWirelessUi.java [hunk 13]
```diff
@@ -423,6 +423,11 @@ public void getUpdatedNetInterface(GwtNetInterfaceConfig updatedNetIf) {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/network/TabWirelessUi.java:426]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/ApplicationCertsTabUi.java [hunk 14]
```diff
@@ -197,4 +197,10 @@ private boolean isAppCertValid() {
             return true;
         }
     }
+
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/ApplicationCertsTabUi.java:201]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/DeviceCertsTabUi.java [hunk 15]
```diff
@@ -231,4 +231,9 @@ private boolean isDeviceCertValid() {
             return true;
         }
     }
+
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/DeviceCertsTabUi.java:235]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/SecurityTabUi.java [hunk 16]
```diff
@@ -92,6 +92,11 @@ public void refresh() {
         }
     }
 
+    @Override
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/SecurityTabUi.java:95]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/ServerCertsTabUi.java [hunk 17]
```diff
@@ -110,6 +110,11 @@ public void refresh() {
         }
     }
 
+    @Override
+    public void clear() {
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/ServerCertsTabUi.java:114]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/SnapshotsTabUi.java [hunk 18]
```diff
@@ -194,6 +194,11 @@ public void onSuccess(ArrayList<GwtSnapshot> result) {
         });
     }
 
+    @Override
+    public void clear() {
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/settings/SnapshotsTabUi.java:198]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/util/TextFieldValidator.java [hunk 19]
```diff
@@ -76,7 +76,7 @@ public void onChange(ChangeEvent event) {
         NETWORK("network", "(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,3})"),
         IPv4_ADDRESS("ipv4_address", "\\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\b"),
         PORT("port", "^[0-9]*$"),
-        PORT_RANGE("port_range", "^[0-9]+:*[0-9]+$"),
+        PORT_RANGE("port_range", "^[0-9]+:[0-9]+$"),
```
Comment (LeoNerdoG) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/util/TextFieldValidator.java:79]: Please change the header to: " * Copyright (c) 2011, 2020 Eurotech and/or its affiliates"

**Review by LeoNerdoG — COMMENTED** [review]
Hey @pierantoniomerlino,
I have found some headers that were not changed, can you please change the headers to "2020"?

**Review by LeoNerdoG — APPROVED** [review]

**Comment by pierantoniomerlino** [comment]
@MMaiero @LeoNerdoG I tried to address more corner cases in Port/IPForwarding tabs. Could you please check again this PR? Try with all sort of operations on the firewall tabs..

**Comment by MMaiero** [comment]
It seems to be working now as expected.
I propose to merge this.

**Review by MMaiero — APPROVED** [review]

**Comment by MMaiero** [comment]
We need to start considering to refactor the tab code in order to have a unified approach throughout all the UI.
Maybe we can open a corresponding issue

**Pull request merged**